### PR TITLE
Implementation of pull-based resources

### PR DESCRIPTION
**Pull request by tcalmant**
* Uses `isExternalGet` and `isExternalSet` flags of the `ResourceAttribute` to check if an external handler must be called
* Adds the `externalGetCacheMs` entry in `ResourceAttribute` to define the time to wait before calling again the external Get handler
* The value returned by the handlers is stored in the twin
* Added `GetLevel` to define 3 levels of resource access:
  * `WEAK`: always return the twin value, never call the external get
  * `NORMAL`: call the external get if no twin value is available or if the `externalGetCacheMs` duration expired
  * `STRONG`: always call the external get.
* Doesn't work with resources defined outside the ModelBuilder, like pre-defined admin resources.

Also adds a small fix in the history provider integration tests to protect against a null pointer when Docker is not available.

### CODE REVIEW

File: prototype/core/annotation/src/main/java/org/eclipse/sensinact/core/annotation/verb/GET.java [hunk 1]
```diff
@@ -73,6 +75,19 @@
...
+     * the cache duration is exceeded, the annotated method will be called.
+     *
+     * The default duration is in {@link TimeUnit#MILLISECONDS}.
+     */
+    long cacheDuration() default 500;
+
+    /**
+     * Unit of the value cache duration. Defaults to {@link ChronoUnit#MILLIS}.
+     */
+    ChronoUnit cacheDurationUnit() default ChronoUnit.MILLIS;
```
Comment (timothyjward) [inline prototype/core/annotation/src/main/java/org/eclipse/sensinact/core/annotation/verb/GET.java:89]: Should this be TimeUnit rather than ChronoUnit? My reasoning is that we're dealing with relative times not dates, so TimeUnit is more appropriate. It also links with the JavaDoc for `cacheDuration` which says `TimeUnit`.

File: prototype/core/api/src/main/java/org/eclipse/sensinact/core/command/GetLevel.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+     * threshold, pull the real value, else return the cached one. For push-based
+     * values, returns the last cached value.
+     */
+    CACHED,
+
+    /**
+     * Weak Get: always return the last value in cache. Don't try to pull a real
+     * value
+     */
+    WEAK,
```
Comment (timothyjward) [inline prototype/core/api/src/main/java/org/eclipse/sensinact/core/command/GetLevel.java:31]: We should document the behaviour if the value has never been set. Would it do a pull in this case? I think probably not, but we should be explicit.

File: prototype/core/api/src/main/java/org/eclipse/sensinact/core/command/GetLevel.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * Hard Get: always ask for the real value and put it in cache. Acts like
+     * {@link #CACHED} for pushed values.
+     */
+    HARD,
```
Comment (timothyjward) [inline prototype/core/api/src/main/java/org/eclipse/sensinact/core/command/GetLevel.java:37]: This is notionally the opposite of a `WEAK` get so I think that this constant name should be `STRONG`, or that `WEAK` should be changed to `SOFT`?

File: prototype/core/api/src/main/java/org/eclipse/sensinact/core/model/ResourceBuilder.java [hunk 4]
```diff
@@ -85,20 +85,21 @@
...
 
     /**
-     * Set a getter function to be called
-     *
-     * @param getter
-     * @return
+     * This resource has a dynamic get behavior. Can't work on resource with an action.
      */
-    ResourceBuilder<B, T> withGetter(Supplier<T> getter);
+    ResourceBuilder<B, T> withGetter();
```
Comment (timothyjward) [inline prototype/core/api/src/main/java/org/eclipse/sensinact/core/model/ResourceBuilder.java:90]: Should we have a boolean version as well so that people can say `withGetter(false)` when copying another model?

File: prototype/core/api/src/main/java/org/eclipse/sensinact/core/model/ResourceBuilder.java [hunk 5]
```diff
@@ -85,20 +85,21 @@
...
-     * @return
+     * @param cacheDuration Duration of the cache
+     */
+    ResourceBuilder<B, T> withGetterCache(Duration cacheDuration);
+
+    /**
+     * This resource has a dynamic set behavior. Can't work on resource with an action.
      */
-    ResourceBuilder<B, T> withSetter(Consumer<T> setter);
+    ResourceBuilder<B, T> withSetter();
```
Comment (timothyjward) [inline prototype/core/api/src/main/java/org/eclipse/sensinact/core/model/ResourceBuilder.java:102]: As with `withGetter` a boolean may be useful?

File: prototype/core/impl/src/main/java/org/eclipse/sensinact/prototype/command/impl/GatewayThreadImpl.java [hunk 6]
```diff
@@ -153,6 +155,41 @@ private Promise<Object> performAction(String model, String provider, String serv
         return d.getPromise();
     }
 
+    private <T> Promise<TimedValue<T>> pullResourceValue(String model, String provider, String service,
```
Comment (timothyjward) [inline prototype/core/impl/src/main/java/org/eclipse/sensinact/prototype/command/impl/GatewayThreadImpl.java:158]: This behaviour should live outside the gateway thread, probably in the model nexus or the resource itself. 
Having reflected, we shouldn't be grabbing the gateway thread's promise factory for the work either as that's really for command use. GET/SET/ACT should share their own thread pool for doing the calls and then update the twin on the gateway thread with a command.

File: prototype/core/impl/src/main/java/org/eclipse/sensinact/prototype/command/impl/GatewayThreadImpl.java [hunk 7]
```diff
@@ -153,6 +155,41 @@ private Promise<Object> performAction(String model, String provider, String serv
...
+                    d.fail(e);
+                }
+            });
+        } catch (Exception e) {
+            d.fail(e);
+        }
+        return d.getPromise();
+    }
+
+    private <T> Promise<TimedValue<T>> pushResourceValue(String model, String provider, String service,
```
Comment (timothyjward) [inline prototype/core/impl/src/main/java/org/eclipse/sensinact/prototype/command/impl/GatewayThreadImpl.java:175]: As with pull, this shouldn't be here. It should be done in the nexus/resource

File: prototype/core/impl/src/main/java/org/eclipse/sensinact/prototype/impl/SensiNactSessionImpl.java [hunk 8]
```diff
@@ -216,8 +216,31 @@ private <T> T safeGetValue(Promise<T> promise) {
...
+    public <T> TimedValue<T> getResourceTimedValue(String provider, String service, String resource, Class<T> clazz) {
+        return safeExecute(new AbstractTwinCommand<TimedValue<T>>() {
+            @Override
+            protected Promise<TimedValue<T>> call(SensinactDigitalTwin model, PromiseFactory pf) {
+                final SensinactResource sensinactResource = model.getResource(provider, service, resource);
+                if (sensinactResource != null) {
+                    if (sensinactResource.getResourceType() == ResourceType.ACTION) {
+                        return pf.resolved(null);
+                    }
+                    // TODO: get the GetLevel from argument
```
Comment (timothyjward) [inline prototype/core/impl/src/main/java/org/eclipse/sensinact/prototype/impl/SensiNactSessionImpl.java:237]: Does this need tidying up?

File: prototype/core/annotation/src/main/java/org/eclipse/sensinact/core/annotation/verb/GET.java (hunk 1 above)
Comment (tcalmant) [reply prototype/core/annotation/src/main/java/org/eclipse/sensinact/core/annotation/verb/GET.java:89]: As we use `Duration` internally, I'd prefer to keep using `ChronoUnit` and update the Javadoc

File: prototype/core/api/src/main/java/org/eclipse/sensinact/core/command/GetLevel.java (hunk 3 above)
Comment (tcalmant) [reply prototype/core/api/src/main/java/org/eclipse/sensinact/core/command/GetLevel.java:37]: As discussed, we'll use `WEAK`, `NORMAL` and `STRONG`
